File: service-b/app/routers/blacklist.py
from typing import List

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from app.database import get_db
from app.deps import get_current_user, require_admin
from app.models import Blacklist, User
from app.schemas import BlacklistCreate, BlacklistOut

router = APIRouter(prefix="/api/v1/blacklist", tags=["Blacklist"])


@router.get("", response_model=List[BlacklistOut])
def list_blacklist(
    db: Session = Depends(get_db),
    _: User = Depends(get_current_user),
):
    return db.query(Blacklist).order_by(Blacklist.added_at.desc()).all()


@router.post("", response_model=BlacklistOut, status_code=201)
def add_to_blacklist(
    payload: BlacklistCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_admin),
):
    existing = db.query(Blacklist).filter(Blacklist.plate_number == payload.plate_number.upper()).first()
    if existing:
        raise HTTPException(status_code=409, detail="Plate already in blacklist")
    entry = Blacklist(
        plate_number=payload.plate_number.upper(),
        reason=payload.reason,
        added_by=current_user.username,
    )
    db.add(entry)
    db.commit()
    db.refresh(entry)
    return entry


@router.delete("/{plate_number}", status_code=204)
def remove_from_blacklist(
    plate_number: str,
    db: Session = Depends(get_db),
    _: User = Depends(require_admin),
):
    entry = db.query(Blacklist).filter(Blacklist.plate_number == plate_number.upper()).first()
    if not entry:
        raise HTTPException(status_code=404, detail="Plate not found in blacklist")
    db.delete(entry)
    db.commit()

File: frontend/src/api/blacklist.js
/**
 * blacklist.js — Blacklist REST API calls.
 */
import client from './client'

/** GET /api/v1/blacklist — all blacklisted plates. */
export async function getBlacklist() {
  const { data } = await client.get('/api/v1/blacklist')
  return data
}

/**
 * POST /api/v1/blacklist — add a plate (admin only).
 * @param {string} plate_number
 * @param {string} reason
 */
export async function addToBlacklist(plate_number, reason) {
  const { data } = await client.post('/api/v1/blacklist', { plate_number, reason })
  return data
}

/**
 * DELETE /api/v1/blacklist/{plate_number} — remove a plate (admin only).
 */
export async function removeFromBlacklist(plateNumber) {
  const { data } = await client.delete(`/api/v1/blacklist/${encodeURIComponent(plateNumber)}`)
  return data
}
